ludwig/data/dataset/tfrecord.py

```diff
@@ -0,0 +1,211 @@
+"""TF record dataset."""
+import contextlib
+import glob
+import json
+import math
+
+import os
+import tensorflow as tf
+from ludwig.constants import NAME, PROC_COLUMN
+from ludwig.data.batcher.iterable import IterableBatcher
+from ludwig.data.dataset.base import Dataset
+from ludwig.data.dataset.pandas import PandasDataset
+from ludwig.data.dataset.partitioned import PartitionedDataset
+from ludwig.utils.data_utils import DATA_TRAIN_HDF5_FP
+from ludwig.utils.fs_utils import to_url
+from ludwig.utils.misc_utils import get_combined_features, get_proc_features
+
+
+class TFRecordDataset(Dataset):
+    def __init__(self, url, features, training_set_metadata):
+        self.url = to_url(url)[7:]
+        if not os.path.isdir(self.url):
+            raise RuntimeError("url for TFRecordDataset must be a folder.")
+        abs_path = os.path.abspath(os.path.expanduser(self.url))
+        self.file_names = [os.path.join(abs_path, name)
+                           for name in glob.glob(os.path.join(abs_path, "*.tfrecords.gz"))]
+        self.features = [feature[PROC_COLUMN] for feature in features]
+        self.training_set_metadata = training_set_metadata
+
+        # dataset = tf.data.TFRecordDataset(self.file_names, compression_type="GZIP")
+        # self.feature, self.feature_lists = self._detect_schema(dataset)
+        with open(os.path.join(abs_path, "meta.json")) as in_file:
```

Comment:
- Unlike petastorm, tfrecord does not have metadata. A typical meta info is `dataset.size`, which can be queried via petastorm API but not tf.data API  (tf.data.cardinality won't work because the dataset is constructed from files). I workaround this issue by writing the size into a json at construction time. See [here: ](https://github.com/zhisbug/ludwig/blob/hao-add-tfrecord/ludwig/data/dataset/tfrecord.py#L32).